uber/models.py

```diff
@@ -494,93 +494,138 @@ def lookup_attendee(self, full_name, email, zip_code):
-            else:
+            old_badge_num = int(old_badge_num or 0) or None
+
+            if old_badge_type == attendee.badge_type and (not attendee.badge_num or old_badge_num == attendee.badge_num):
+                attendee.badge_num = old_badge_num
+                return 'Attendee is already {} with badge {}'.format(c.BADGES[old_badge_type], old_badge_num)
+
+            if c.SHIFT_CUSTOM_BADGES:
                 # fill in the gap from the old number, if applicable
+                badge_num_keep = attendee.badge_num
                 if old_badge_num:
                     self.shift_badges(old_badge_type, old_badge_num + 1, down=True)
 
-                # determine the new badge number now that the badges have shifted
-                next = self.next_badge_num(badge_type, old_badge_num)
-                badge_num = min(int(badge_num) or next, next)
-
                 # make room for the new number, if applicable
-                if badge_num:
-                    offset = 1 if badge_type == old_badge_type and old_badge_num and badge_num > old_badge_num else 0
-                    self.shift_badges(badge_type, badge_num + offset, up=True)
+                if attendee.badge_num:
+                    offset = 1 if old_badge_type == attendee.badge_type and attendee.badge_num > (old_badge_num or 0) else 0
+                    self.shift_badges(attendee.badge_type, attendee.badge_num + offset, up=True)
+                attendee.badge_num = badge_num_keep
+
+            if not attendee.badge_num and self.needs_badge_num(attendee):
+                attendee.badge_num = self.get_next_badge_num(attendee.badge_type)
 
-            attendee.badge_num = badge_num
-            attendee.badge_type = badge_type
             return 'Badge updated'
 
+        def auto_badge_num(self, badge_type):
+            """
+            Gets the next available badge number for a badge type's range.
+
+            This ignores the session so that plugins can override the logic here if need be without worrying about
+            handling dirty sessions.
+
+            :param badge_type: Used as a starting point if no badges of the same type exist, and to select badges within
+            a specific range.
+            :return:
+            """
+            sametype = self.query(Attendee.badge_num).filter(Attendee.badge_type == badge_type,
+                                                   Attendee.badge_num >= c.BADGE_RANGES[badge_type][0],
+                                                   Attendee.badge_num <= c.BADGE_RANGES[badge_type][1])
+            if sametype.count():
+                sametype_list = [int(row[0]) for row in sametype.order_by(Attendee.badge_num).all()]
+
+                # Searches badge range for a gap in badge numbers; if none found, returns the latest badge number + 1
+                first = 0
+                last = sametype.count() - 1
+                if sametype_list[last] - sametype_list[first] == last:
```

Comment:
This extra if/else statement theoretically shouldn't be needed, but I hit a wall when figuring out what in the binary sort was breaking the code - none of the unit tests would run because all the pre-set staff badges were being given `None` as a badge number.